content/blog/client-futures.md

```diff
@@ -0,0 +1,70 @@
+
+OK, back to the futures migration. Whenever you end up using unsafe methods on `Pin` you have to think about all of the above in order to ensure the pinning invariants. See the [std::pin docs](https://doc.rust-lang.org/nightly/std/pin/index.html) for more on how to reason about this. In many of the places I had to use field projection, the reason is to call `poll` on another future (sometimes indirectly). For that you need a pinned pointer and so you need *structural pinning* (i.e., you project a field of `Pin<&mut T>` to `Pin<&mut FieldType>`).
+
+## Functions
+
+An irritating aspect of the migration was that a lot of the names of functions (and types) in the futures library have changed. These name changes were sometimes hard to automate because the names are often shared with very common names in standard libraries. Some example name changes are `Async` to `Poll`, `Ok` to `ready`, `for_each` to `then`, `then` to `map`, and `Either::A` to `Either::Left` (likewise, `B`/`Right`).
+
+Sometimes the name stayed the same, but the semantics of the function changed slightly (or both!). A common change was taking a closure which returned a future that produces a `T` rather than returning a `T` itself.
+
+Many of the combinator functions moved from the `Future` trait to one of the extension crates. This is an easy issue to fix, but sometimes this fix is hard to deduce from the error message.
+
+
+## LoopFn
+
+The futures 0.1 library included a `LoopFn` future for handling futures which do something multiple times. It has been removed from futures 0.3, which, I believe, is because a `for` loop in an `async` function or using streams are better solutions in the long term. To make the migration straightforward, I created our own version of the 0.3 `LoopFn` future. This was mostly a copy/paste job with some adjustments (e.g., to handle pinning projection): [code](https://github.com/tikv/client-rust/pull/41/commits/6353dbcfe391d66714686aafab9a49e593259dfb#diff-eeffc045326f81d4c46c22f225d3df90R28). Later I converted several uses of `LoopFn` to use streams, which improved the code somewhat.
+
+
+## Sink::send_all
+
+In a few places in the project, we use sinks. I found the migration of these much less easy than for futures. The trickiest issue was that `Sink::send_all` has changed. In 0.1, it took ownership of a stream and once everything is resolved, returned the sink and stream. In 0.3, it takes a mutable reference to the stream and returns nothing. I created our own [compatibility layer](https://github.com/tikv/client-rust/pull/41/commits/6353dbcfe391d66714686aafab9a49e593259dfb#diff-eeffc045326f81d4c46c22f225d3df90R68) to emulate 0.1 sinks with 0.3 futures. This wasn't particularly difficult, but there might be a better way to do this.
```

Comment:
Do you think a  short closing sentence with a link to the PR would help wrap up the post?